### PR TITLE
Refactor to use Vite

### PR DESCRIPTION
**Pull request by Fernthedev**
🦀🌮
Should extensively test

### CODE REVIEW

**Comment by Fernthedev** [comment]
I was curious to see if Vite would offer any performance or size improvements over the current webpack bundler, so I ran some very simple and rudimentary benchmarks (without testing the frontend result itself, keep that in mind) and using the same directory configuration (so basically, served at `/`)

These are the size result comparisons:
| Bundler|  Size | Build Times |
|-------------|-----------|------------|
| Vite | image | 10 seconds with TSC validation |
| React Create Scripts (webpack) | image | 65 seconds |

These are the build logs: 


<details>
 <summary> Vite </summary>

```ts
❯ yarn build
yarn run v1.22.18
$ tsc && vite build
vite v2.9.8 building for production...
✓ 203 modules transformed.
dist/assets/bootstrap-icons.c874e14c.woff2   100.13 KiB
dist/assets/bootstrap-icons.92f8082b.woff    133.91 KiB
dist/index.html                              7.76 KiB
dist/assets/Developers.9bbad543.js           0.85 KiB / gzip: 0.45 KiB
dist/assets/Contributions.0b98d1dd.js        2.55 KiB / gzip: 0.68 KiB
dist/assets/Home.5d57d70b.js                 0.43 KiB / gzip: 0.31 KiB
dist/assets/Models.cc23b954.js               6.45 KiB / gzip: 2.15 KiB
dist/assets/GQLClient.ea88e227.js            20.68 KiB / gzip: 7.48 KiB
dist/assets/Model.f4ec2976.js                3.20 KiB / gzip: 1.27 KiB
dist/assets/GetTumbnail.c52d2a80.js          8.15 KiB / gzip: 3.71 KiB
dist/assets/Login.ce78bff1.js                0.96 KiB / gzip: 0.56 KiB
dist/assets/Logout.9c851669.js               0.24 KiB / gzip: 0.21 KiB
dist/assets/Models.5a860a4e.css              0.05 KiB / gzip: 0.07 KiB
dist/assets/urql.es.061cb8bd.js              59.62 KiB / gzip: 17.04 KiB
dist/assets/index.be07df17.css               220.83 KiB / gzip: 33.49 KiB
dist/assets/index.15b349a1.js                350.53 KiB / gzip: 110.04 KiB
Done in 10.93s.
```
</details> 

<details>
<summary>React Create Scripts (webpack) </summary>

```ts
❯ yarn build
yarn run v1.22.18
$ react-scripts build
Creating an optimized production build...
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <4.6.0

YOUR TYPESCRIPT VERSION: 4.6.3

Please only submit bug reports when using the officially supported version.

=============
Compiled successfully.

File sizes after gzip:

  134.81 kB  build\static\js\main.1029d420.js
  34.96 kB   build\static\css\main.6c9c8930.css
  18.92 kB   build\static\js\726.c4bc89f0.chunk.js
  7.3 kB     build\static\js\391.2b705017.chunk.js
  3.86 kB    build\static\js\751.60661d2c.chunk.js
  2.4 kB     build\static\js\293.7c6d9df3.chunk.js
  2.24 kB    build\static\js\191.f039a887.chunk.js
  1.26 kB    build\static\js\864.9b798b51.chunk.js
  940 B      build\static\js\107.c5463851.chunk.js
  894 B      build\static\js\986.2b433a5c.chunk.js
  641 B      build\static\js\432.f611c5e9.chunk.js
  426 B      build\static\js\831.3f8cf387.chunk.js
  329 B      build\static\js\984.a23ea2da.chunk.js
  120 B      build\static\css\191.514ae7c3.chunk.css
  120 B      build\static\css\293.514ae7c3.chunk.css

The project was built assuming it is hosted at /.
You can control this with the homepage field in your package.json.

The build folder is ready to be deployed.
You may serve it with a static server:

  yarn global add serve
  serve -s build

Find out more about deployment here:

  https://cra.link/deployment

Done in 65.16s.
```
</details>


In this very simple comparison, I would believe Vite is the clear winner both for the developer UX and the production result. Vite offers near instant hot reload and startup, allowing a much better experience working with the React UI. Configuration is also much more concise and well documented compared to webpack (which is notoriously annoying to configure)

Overall, it is my opinion that ModelSaber benefits from this PR in all aspects and very little reason to choose webpack over Vite. Though keep in mind this is naive as I have **not** validated the `build` works as expected.

**Comment by wolfcomp** [comment]
Has this been tested with the ASP.NET web server?

**Comment by Fernthedev** [comment]
It has not been tested with ASP.NET, though I can if instructions were to be provided. I did test with `serve` though and it worked _seemingly_ fine if that works too.

**Comment by raineio** [comment]
Most of the ASP.NET serving instructions are provided through the [ModelSaber](https://github.com/modelsaber/modelsaber) solution repo. From there you should be able to set up both the ASP.NET and NodeJS instances and test properly.

**Comment by Fernthedev** [comment]
I finally got the backend started and working with the frontend. It required minimal changes though I should mention I was stuck in the loading screen and refreshing repeatedly 
image

Some other things I want to mention:
- Using docker for the PostgreSQL instance
  - `docker run --name postgres -e POSTGRES_PASSWORD=postgres -p 5432:5432 -d postgres:alpine`  
  - Please uses Docker compose for development purposes testing alongside ModelSaber
- Lack of development setup documentation made this far more tedious than it needed to be
- Please mention environment variables OR config for database configuration!
- I drank a lot of water and touched lots of grass 

I don't know where next to test for the purposes of this PR, as ultimately you and the ModelSaber team as a whole will be the ones to reap the benefits and suffer the consequences, not me.

File: ClientApp/.gitignore [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# Editor directories and files
+.vscode/*
+!.vscode/extensions.json
+.idea
+.DS_Store
+*.suo
+*.ntvs*
+*.njsproj
+*.sln
+*.sw?
```
Comment (wolfcomp) [inline ClientApp/.gitignore:24]: Should possibly be moved to the main folder for the repo instead of the sub folder.

**Review by wolfcomp — CHANGES_REQUESTED** [review]
The only thing is possibly look at [ClientApp/src/index.tsx](https://github.com/ModelSaber/ModelSaber.Main/blob/master/ClientApp/src/index.tsx#L14) to fix the `process is not defined` error as that is something i know `react-scripts` handles for changing between dev and prod

File: ClientApp/src/index.tsx [hunk 2]
```diff
@@ -11,7 +11,8 @@ const GQLClient = lazy(() => import("./components/GQLClient"));
 const baseUrl = document.getElementsByTagName("base")[0].getAttribute("href");
 const rootElement = document.getElementById("root");
 const loaderbackground = document.getElementById("lds-roller");
-const uri = process.env.NODE_ENV == "development" ? process.env.REACT_APP_API_URL : "https://apimodelsaber.rainemods.io";
+const uri = import.meta.env.DEV ? import.meta.env.VITE_APP_API_URL : "https://apimodelsaber.rainemods.io";
+console.log(`ENV ${JSON.stringify(import.meta.env)}`)
```
Comment (wolfcomp) [inline ClientApp/src/index.tsx:15]: Remove the log so it's not forgotten.

**Review by wolfcomp — CHANGES_REQUESTED** [review]

**Review by wolfcomp — APPROVED** [review]

**Pull request merged**